plugins/slack_app/test/controllers/slack_app_controller_test.rb

```diff
@@ -0,0 +1,100 @@
+require_relative '../test_helper'
+
+SingleCov.covered!
+
+def post_as(id, params)
+  opts = params.merge(
+    token: nil,
+    user_id: id.identifier,
+    response_url: 'http://example.com/blah'
+  )
+  post :command, opts
+end
+
+describe SlackAppController do
+  let(:user1) { users(:deployer) }
+  let(:user2) { users(:deployer_buddy) }
+  let(:user1_identifier) { slack_identifiers(:user1) }
+  let(:user2_identifier) { slack_identifiers(:user2) }
+  before do
+    user1_identifier.update_column(:user_id, user1.id)
+    user2_identifier.update_column(:user_id, user2.id)
+  end
+
+  as_a_viewer do
+    describe '#oauth' do
+      with_env SLACK_CLIENT_ID: 'client-id', SLACK_CLIENT_SECRET: 'client-secret' do
+        it 'sends the user to Slack to connect the app' do
+          get :oauth
+          assert_match 'https://slack.com/oauth/authorize?client_id=', @response.body
+          assert_match '/slack_app/oauth', @response.body
+          assert_match 'chat:write:bot,commands,identify', @response.body
+        end
+
+        it 'accepts an app token from Slack' do
+          stub_request(:post, "https://slack.com/api/oauth.access").
+            to_return(status: 200, body: '{"access_token": "iamatoken", "user_id": "Uabcdef"}')
+          get :oauth, code: 'iamaslackcode'
+          assert SlackIdentifier.app_token
+          identifier = SlackIdentifier.find_by_user_id users(:viewer).id
+          assert_equal identifier.identifier, 'Uabcdef'
+        end
+
+        it 'accepts a user token from Slack' do
+          SlackIdentifier.create! identifier: 'xoxz-i-am-an-app-token'
+          stub_request(:post, "https://slack.com/api/oauth.access").
+            to_return(status: 200, body: '{"access_token": "iamatoken", "user_id": "Udefghi"}')
+          get :oauth, code: 'iamaslackcode'
+          identifier = SlackIdentifier.find_by_user_id users(:viewer).id
+          assert_equal identifier.identifier, 'Udefghi'
+        end
+      end
+    end
+  end
+
+  describe '#command' do
+    it "raises if secret token doesn't match" do
+      assert_raises RuntimeError do
```

Comment:
errors can happen for all kinds of reasons ... make sure it's for the right one :)
e = assert_raises
e.message.must_include "xyz"